pkg/cf/cf.go

```diff
@@ -0,0 +1,567 @@
+				zoneLogger.Debugf("Deleting worker route with ID %s", route.ID)
+				_, err := m.api.DeleteWorkerRoute(m.Ctx, cf.ZoneIdentifier(zone.ID), route.ID)
+				if err != nil {
+					return err
+				}
+				zoneLogger.Debugf("Done deleting worker route with ID %s", route.ID)
+			}
+		}
+	}
+
+	m.logger.Debugf("Listing worker KV Namespaces")
+	kvNamespaces, _, err := m.api.ListWorkersKVNamespaces(m.Ctx, cf.AccountIdentifier(m.AccountCfg.ID), cf.ListWorkersKVNamespacesParams{})
+	if err != nil {
+		return err
+	}
+	m.logger.Tracef("kvNamespaces: %+v", kvNamespaces)
+	m.logger.Debugf("Done listing worker KV Namespaces")
+
+	for _, kvNamespace := range kvNamespaces {
+		if kvNamespace.Title == KVNsName {
+			m.logger.Debugf("Deleting worker KV Namespace with ID %s", kvNamespace.ID)
+			_, err := m.api.DeleteWorkersKVNamespace(m.Ctx, cf.AccountIdentifier(m.AccountCfg.ID), kvNamespace.ID)
+			if err != nil {
+				return err
+			}
+			m.logger.Debugf("Done deleting worker KV Namespace with ID %s", kvNamespace.ID)
+		}
+	}
+
+	m.logger.Debugf("Attempting to delete worker script %s", ScriptName)
+	err = m.api.DeleteWorker(m.Ctx, cf.AccountIdentifier(m.AccountCfg.ID), cf.DeleteWorkerParams{
+		ScriptName: ScriptName,
+	})
+	if err != nil {
+		if !strings.Contains(err.Error(), "workers.api.error.script_not_found") {
+			return err
+		}
+		m.logger.Debugf("Didn't find worker script %s", ScriptName)
+	} else {
+		m.logger.Debugf("Deleted worker script %s", ScriptName)
+	}
+	m.logger.Info("Done cleaning up existing workers")
+	return nil
+}
+
+func (m *CloudflareAccountManager) ProcessDeletedDecisions(decisions []*models.Decision) error {
+	keysToDelete := make([]string, 0)
+	newKVPairByValue := make(map[string]cf.WorkersKVPair)
+	for _, kvPair := range m.KVPairByDecisionValue {
+		newKVPairByValue[kvPair.Value] = kvPair
+	}
+
+	for _, decision := range decisions {
+		if *decision.Scope == "range" {
+			delete(m.ActionByIPRange, *decision.Value)
+			continue
+		}
+		if val, ok := m.KVPairByDecisionValue[*decision.Value]; ok {
+			if *decision.Type == val.Value {
+				keysToDelete = append(keysToDelete, val.Key)
+				delete(newKVPairByValue, val.Value)
+			}
+		}
+	}
+	if len(keysToDelete) == 0 {
+		m.logger.Debug("No keys to delete")
+		return nil
+	}
+	m.logger.Infof("Deleting %d decisions", len(keysToDelete))
+	deleterGrp := errgroup.Group{}
+	for batch, i := 0, 0; i < len(keysToDelete); i += 10000 {
+		batch++
+		batch := batch
+		begin := i
+		end := min(i+10000, len(keysToDelete))
+		deleterGrp.Go(func() error {
+			resp, err := m.api.DeleteWorkersKVEntries(m.Ctx, cf.AccountIdentifier(m.AccountCfg.ID), cf.DeleteWorkersKVEntriesParams{
+				Keys:        keysToDelete[begin:end],
+				NamespaceID: m.NamespaceID,
+			})
+			if err != nil {
+				return err
+			}
+			m.logger.Tracef("batch %d delete key resp: %+v", batch, resp)
+			return nil
+		})
+	}
+	if err := deleterGrp.Wait(); err != nil {
+		return err
+	}
+	m.logger.Info("Done deleting keys")
```

Comment:
can we log how many decisions we deleted ? and have a better log eg "%d decisions have been deleted"